File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int buy= 0;
        int sell = buy+1;
        int max_profit=0;
        
        while(buy<n-1 && sell<n){            
                if(prices[sell]>prices[buy]){
                    int current_profit= prices[sell]-prices[buy];
                    max_profit = Math.max(current_profit, max_profit);
                }
            else {
                buy = sell; // Update the buying index
            }
                
            sell++;
            }
        
        return max_profit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        left, right = 0, 1
        max_profit = 0
        
        while right<len(prices):
            curr_profit = prices[right]-prices[left]
            if curr_profit < 0:
                left = right
                right = left+1
            else:
                right+=1                
                max_profit = max(curr_profit, max_profit)
                
        return max_profit
        